File: python/questions/rotateRight.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head == None:
            return None
        ln = 1
        cur = head
        while cur.next:
            cur = cur.next
            ln += 1
        k = ln - k % ln
        cur.next = head
        while k > 0:
            k -= 1
            cur = cur.next
        head = cur.next
        cur.next = None
        return head

File: cpp/questions/rotateRight.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == nullptr) {
            return nullptr;
        }
        int ln = 1;
        ListNode* cur = head;
        while (cur->next) {
            cur = cur->next;
            ln++;
        }
        k = ln - k % ln;
        cur->next = head;
        while (k-- > 0) {
            cur = cur->next;
        }
        head = cur->next;
        cur->next = nullptr;
        return head;
    }
};
